include/init.h

```diff
@@ -25,49 +26,98 @@ extern "C" {
 #define _SYS_INIT_LEVEL_POST_KERNEL	2
 #define _SYS_INIT_LEVEL_APPLICATION	3
 
+struct device;
+
+/**
+ * @brief Static init entry structure for each device driver or services
+ *
+ * @param dev pointer to a device driver instance structure. Can be NULL
+ * if this init entry is not used for a device driver.
+ * @param init init function for the init entry. Also used to initialize
+ * the device driver instance represented by
+ */
+struct init_entry {
+	int (*init)(struct device *dev);
+	struct device *dev;
```

Comment:
Seems like this could be moved into device.h entirely, if it is really device specific. Any point to keep it here?

Reply:
Because init_entry stuff is not device specific, actually. It's "generic". Note that, if required, in future we could have a union {} around the dev, and a marker about the type on the init. (Though I don't any requirement for that yet. But it's possible)

Reply:
Then why not just have this be a `void *arg` and be done with it?

Reply:
I don't want to do casting on such low level parts. First because it will not give you any gain from memory point of view and is less readable. Plus, next phase (with the device status reporting stuff) will be to "const-ify" all struct device everywhere, with all that implies.

Reply:
I get the point about "less readable", but casting a `struct device *` to a `struct notdevice *` violates MISRA 11.3, which is a required constraint, compared to 11.4 which only advises against casting a `void *` to a `struct device *`.  So maybe it really is better as:
```
struct device_init_entry {
    int (*init)(struct void *p);
    void *p;
};
```
But I'm not sure if the future constifying makes that irrelevant.

Reply:
I propose that this issue is resolved by the previous comment block and documentation that a null device pointer field/argument indicates that the init function does not take any arguments.